src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/CodeActions/Razor/WrapAttributesCodeActionProvider.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Models;
+using Microsoft.CodeAnalysis.Razor.CodeActions.Razor;
+using Microsoft.CodeAnalysis.Razor.Protocol;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Razor.CodeActions;
+
+internal class WrapAttributesCodeActionProvider : IRazorCodeActionProvider
+{
+    public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context.HasSelection)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var owner = syntaxTree.Root.FindNode(TextSpan.FromBounds(context.StartAbsoluteIndex, context.EndAbsoluteIndex));
+        if (owner is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var attributes = FindAttributes(owner);
+        if (attributes.Count == 0)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var first = true;
+        var firstAttributeLine = 0;
+        var indentSize = 0;
+        var sourceText = context.SourceText;
+
+        using var newLinePositions = new PooledArrayBuilder<int>();
+        foreach (var attribute in attributes)
+        {
+            var linePositionSpan = attribute.GetLinePositionSpan(context.CodeDocument.Source);
+
+            if (first)
+            {
+                firstAttributeLine = linePositionSpan.Start.Line;
+                sourceText.TryGetFirstNonWhitespaceOffset(attribute.Span, out var indentSizeOffset);
```

Comment:
>TryGetFirstNonWhitespaceOffset [](http://example.com/codeflow?start=27&length=30)

Are there cases when attribute span starts with whitespace?

Reply:
Yes, the Razor syntax tree continues to be annoying, so all of the attributes start with whitespace. The syntax nodes look a bit like`[<][foo][ class="asdf"][/>]`.